[WIP] Cleaned the include headers in core/

santouits opened the pull request:
Hi, I cleaned the headers in the core folder and made forward declarations when possible.

I also moved the defining of DEBUG_METHODS_ENABLED from the source code to the detect.py inside every platform's folder. I think it's better these things to be handled from the build system. Maybe it should also be removed and just leave the DEBUG_ENABLED but that's a different discussion.

I compiled for windows, linux, javascript, and travis passes it's builds also, but there are so many different configurations and there are some that may need some headers to be included, so I'll list the most requested headers I encountered, so you search them first if your compiler complains about a type missing.
```
core/print_string.h // if your compiler can't find print_error and other print commands
core/engine.h
core/os/os.h
core/script_language.h // if your compiler can't find Script, ScriptInstance, get_script_instance()
core/image.h
core/os/main_loop.h
core/os/file_access.h
core/int_types.h
core/typedefs.h
core/math/math_defs.h
```


Also I have a problem, maybe @neikeq can help. I get this error when opening the editor:
```
ERROR: GetTypeInfo<class MainLoop *,void>::get_class_info: GetTypeInfo
    fallback. Bug!
```

neikeq commented:
The `GetTypeInfo` error is very likely because you're forward declaring `MainLoop`. `GetTypeInfo<MainLoop *>` should match the template spetialization `struct GetTypeInfo<const T *, typename EnableIf<TypeInherits<Object, T>::value>::type>`. However, if `MainLoop` is forward declared, there is no way for `GetTypeInfo` to know it inherits `Object`.

neikeq commented:
I can confirm that adding `#include "core/os/main_loop.h"` to `core/bind/core_bind.h` fixes it.

neikeq commented:
Actually, including main loop in `core_bind.cpp` also fixes it, so you don't need to include it in the header which is what this PR is trying to avoid.

neikeq commented:
Also, I had to add `#include "core/print_string.h"` to [modules/mono/glue/gd_glue.cpp](https://github.com/godotengine/godot/blob/master/modules/mono/glue/gd_glue.cpp) when building without `mono_glue=no`.

akien-mga commented on core/class_db.cpp:
```diff
@@ -359,6 +358,8 @@ uint64_t ClassDB::get_api_hash(APIType p_api) {
 	OBJTYPE_RLOCK;
 #ifdef DEBUG_METHODS_ENABLED
 
+#include "core/version.h"
```
I'm not sure about this change, I prefer to have all includes at the top. If needed, you can add a trailing comment specifying what the include is used for.

I haven't seen many C or C++ codebases using such "include right before the use" logic, and while it's pretty common in Python, it also leads to including things several times in different scopes.

santouits replied:
It's alright I will check for debug_method_enabled in the head of the file again

akien-mga commented:
> I also moved the defining of DEBUG_METHODS_ENABLED from the source code to the detect.py inside every platform's folder. I think it's better these things to be handled from the build system. Maybe it should also be removed and just leave the DEBUG_ENABLED but that's a different discussion.

What do we gain from that apart from longer compiler calls? If it was enabled inconditionally as long as `DEBUG_ENABLED` is on, there's no reason to make it a platform-specific define. It adds to the risk of forgetting it in one platform/a thirdparty port and have things break.

I agree that we could drop the define completely and use `DEBUG_ENABLED` only indeed, but that should be done and discussed in a separate PR, so I'm not sure it's wise to do this move to platform code here.

akien-mga commented:
Could you document how you proceeded to clean the includes? Did you use [include-what-you-use](https://include-what-you-use.org/) or another method?

Since we don't have checks in place to prevent over-including in the future, it's important that we know how to proceed to clean things up periodically -- at least until we find a way to prevent such issues from occurring.

santouits commented:
> What do we gain from that apart from longer compiler calls? If it was enabled inconditionally as long as DEBUG_ENABLED is on, there's no reason to make it a platform-specific define. It adds to the risk of forgetting it in one platform/a thirdparty port and have things break.

I agree we shouldn't make it platform specific, I just saw where DEBUG_ENABLED was defined and it was defined in platform and I added it there. We should define it in SConstruct for all platforms I guess.
My thinking for removing it from the source file was the same as you, what happens if someone forgets to include method_bind.h, or if it is removed from another header that someone includes in his file and they think it is still inside, but it fails silently. 
Ok I will reverse that change and search for every occasion that is needed and check if the header is added, and in another PR we can discuss how it will be done

> Could you document how you proceeded to clean the includes? Did you use include-what-you-use or another method?

As I can tell with forward declaration you tell the compiler it's ok trust me I have this type in another file that you will get at linking time, you don't need it anyway, pointers are the same size. That means if a class had something like this:
```
class RandomClass {
    SomeType *some_type;
};
```
we can just get away with it by just writing
```
class SomeType; // trust me it exists
```
but if the type is not a pointer like this:
```
class RandomClass {
    SomeType some_type;
};
``` 
we have to include the header for reasons.
In cpp files though you have to add the header that has the type because you use it like this and it needs to see the signature of the class or the method or something:
```
#include "some_type.h"
some_type->some_method(some_argument);
```
My method was just read the headers in core/ and forward declare everything that was forward declarable like pointers, and as I found later template arguments too. Then I added to the cpp file the header that was needed. I didn't use any tool. I just compiled and when the compiler wanted something I fed it. 
I also didn't touch any code, I'm sure more stuff could be forward declarable if we change some arguments in functions to be references and not copies of the objects, as it is done in many files.

As of having a style for forward declares, I think it is simple, forward declare everything that can be done in the header files. The link you sent has some arguments https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md

santouits committed:
Cleaned the included headers in core/

by forward declaring anything that could be forward declared

akien-mga commented:
Any chance to rebase this? When it is, ping me so that it can be merged ASAP before new merge conflicts arise.

santouits commented:
It is not ready to be merged yet, I have some crashes when building with mono. I have to find first if it is my setup or this code.

neikeq commented:
@santouits I didn't experience any problems. If you want to be sure, send me the backtrace and mono log.

aaronfranke commented:
@santouits Is this still desired? If so, it needs to be rebased on the latest master branch.

akien-mga commented:
Closing as it's not mergeable in this state, and would need to be redone from scratch.
I also attempted to do something similar with `include-what-you-use` a few times but it leads to various issues due either to tooling mistakes or the risk of cyclic dependencies that we have in core around `object.h`, `method_bind.h`, `class_db.h`, etc.
